Fix topdown

cryptoAtwill opened the pull request:
Contain fixes and updates from integration testing. Currently the happy path from this branch is working. Key changes:
- When top down messages are queried, we need to query the next block hash for lotus. See sync for more comments.
- Mint to gateway when cross messages are executed.
- Moving as much code as possible to `GatewayCaller`.

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -238,48 +246,31 @@ where
+                    let validator_changes = provider
+                        .validator_changes_from(prev_height + 1, p.height as u64)
+                        .await?;
+                    let (state, _, _) = self
```
The ignored variables are bothering me a little. Is one of them the `ret`? We should not ignore errors here. For now we should stick to returning an overall error if any of these fail.

aakoshh replied:
Right, I see that inside these methods they return errors. It's a bit confusing that everything is ignored here. 

Also, shouldn't we take at least the events from these and attach them to the final returned result?

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -304,6 +295,90 @@ where
+            .deliver(state, VerifiableMessage::NotVerify(msg))
+            .await?;
+        if is_error(&ret) {
+            return Err(anyhow!("cannot commit parent finality"));
```
It would be nice to attach some info about the error.

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -238,48 +246,31 @@ where
-                    if ret.is_err() {
-                        return Err(anyhow!("failed to apply cross messages"));
-                    }
+                    let (state, ret, _) = self.execute_topdown_msgs(state, messages).await?;
```
```suggestion
                    let (state, ret, ()) = self.execute_topdown_msgs(state, messages).await?;
```
Makes it clear what the return value is.

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -238,48 +246,31 @@ where
+                    let validator_changes = provider
+                        .validator_changes_from(prev_height + 1, p.height as u64)
+                        .await?;
+                    let (state, _, _) = self
```
```suggestion
                    let (state, _ret, ()) = self
```
To make it clear that `()` is returned, and if not it wouldn't compile, so it can't be ignored.

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -420,22 +495,24 @@ fn relayed_bottom_up_ckpt_to_fvm(
+        (provider.genesis_epoch()?, None)
+    } else {
+        (prev_finality.height, Some(prev_finality))
+    })
```
I think it would be good to move most of these into a `topdown` module, similar to the `checkpoint` module. They are so numerous and complicated, they completely dominate the `chain.rs` file by now.

aakoshh commented on fendermint/vm/interpreter/src/fvm/state/ipc.rs:
```diff
@@ -202,6 +204,34 @@ impl<DB: Blockstore> GatewayCaller<DB> {
+            .deserialize::<BytesDe>()
+            .context("failed to deserialize return data")?;
+
+        let function = router::GATEWAYROUTERFACET_ABI
```
```suggestion
        let function = self.router.contract().abi()
```

aakoshh commented on fendermint/vm/interpreter/src/fvm/state/ipc.rs:
```diff
@@ -202,6 +204,34 @@ impl<DB: Blockstore> GatewayCaller<DB> {
+        let function = router::GATEWAYROUTERFACET_ABI
+            .functions
+            .get("commitParentFinality")
+            .ok_or_else(|| anyhow!("broken abi"))?
+            .get(0)
+            .ok_or_else(|| anyhow!("function not found, abi wrong?"))?;
```
Would be good to at least mention what it was looking for.

aakoshh commented on fendermint/vm/interpreter/src/fvm/state/ipc.rs:
```diff
@@ -202,6 +204,34 @@ impl<DB: Blockstore> GatewayCaller<DB> {
+            .get(0)
+            .ok_or_else(|| anyhow!("function not found, abi wrong?"))?;
+
+        if let Token::Tuple(mut tuple) = decode_function_data(function, return_data.0, false)? {
```
`decode_function_data` is generic. I'm sure we can decode into the expected type instead of dealing with popping `Token`s.

aakoshh replied:
```suggestion
        match decode_function_data::<(bool, router::ParentFinality)>(function, return_data.0, false)? {
```

aakoshh reviewed (commented):
Impressive! :clap: 

Do you handle temporary network failures when looking up stuff from the parent? I can't remember where the backoff was applied.

cryptoAtwill commented:
@aakoshh There is a exponential back off implemented [here](https://github.com/consensus-shipyard/fendermint/blob/0d584fe33fa02da088f3b2ed06fde3f8514d7f8a/fendermint/vm/topdown/src/finality.rs#L98).

Some key follow up changes:
- Added `topdown` module and shifted top down parent finality execution logic to `topdown` module.
- Added `call_with_return` method in `ContractCaller`. This way we can obtain the `FvmApplyRet` from calling the contracts.
- Added `ChainMessageApplyRet::Ipc(FvmApplyRet)`, so that we dont have to used `Signed` for unsigned IPC messages.

Currently for top down execution result, only the `topdown` cross message execution apply result is returned. This is because:
- Merging `ApplyRet` seems difficult and awkward to do.
- Will merge `commitParentFinality` and `storeValidatorChanges` into one method in the future, so we can do the ret merging there.
- Currently if we dont return the `ApplyRet` from `commitParentFanlity` and `storeValidatorChanges` is still acceptable for UX (I could be wrong), but good to have for debugging. So for this iteration, I think it is still ok.

aakoshh commented on fendermint/vm/interpreter/src/chain.rs:
```diff
@@ -56,6 +58,8 @@ pub struct IllegalMessage;
 pub enum ChainMessageApplyRet {
     Signed(SignedMessageApplyRes),
+    /// The IPC chain message execution result
+    Ipc(FvmApplyRet),
```
I agree this is cleaner!

aakoshh commented on fendermint/vm/interpreter/src/fvm/state/ipc.rs:
```diff
@@ -189,54 +186,65 @@ impl<DB: Blockstore> GatewayCaller<DB> {
+        let messages = cross_messages
+            .into_iter()
+            .map(router::CrossMsg::try_from)
+            .collect::<Result<Vec<_>, _>>()?;
```
```suggestion
            .collect::<Result<Vec<_>, _>>()
            .context("failed to convert cross messages")?;
```

I hope this cannot be triggered by malicious users.

aakoshh reviewed (approved):
Great! I also couldn't really find a justification for calling the inner interpreter with the the `NotValidate` kind and then trying to decode the return value when the context has been lost. Doing more in the `GatewayCaller` is much simpler :+1:
